docs/proposal/2022-11-22-botkube-plugins-integration.md

```diff
@@ -0,0 +1,498 @@
+    jsonSchema:
+      value: |-
+        {
+          "$schema": "http://json-schema.org/draft-07/schema",
+          "type": "object",
+          "required": [
+              "version"
+          ],
+          "properties": {
+            "version": {
+              "$id": "#/properties/version",
+              "type": "string",
+              "minLength": 5,
+              "pattern": "^(?:0|[1-9]\\d*)\\.(?:0|[1-9]\\d*)\\.(?:0|[1-9]\\d*)$",
+              "title": "Kubernetes version",
+              "description": "Kubernetes version",
+              "default": "",
+              "examples": [
+                  "1.19.0"
+              ]
+            }
+          },
+          "additionalProperties": true
+        }
+```
+
+This will be used by Botkube to validate that configuration defined by user is valid before even starting a given plugin.
+
+### Passing configuration to a given plugin
+
+Current Botkube implementation allows you to specify different executor and sources configuration and bind them to a single channel. Later they are merged together based on the given business logic.
+
+To support the same experience for plugins, we need to pass a list of configuration each time we run the:
+
+- `Execute` method for executor plugins
+- `Source` method for source plugins
+
+We pass an array of strings, so later they can be unmarshalled by a given plugin and merged based on custom business logic.
+
+We don't support configuration validation inside Botkube. See validating [plugins configuration](#plugin-configuration) for more information.
+
+The allowed configuration parameters will be described on the `docs.botkube.io` site in the same way that we did for `kubectl` executor and `kuberneetes` source.
+
+### Refreshing already loaded plugins
+
+Once the Botkube starts it caches:
+
+- all index files for defined plugins repositories
+- all enabled plugins binaries
+
+We use the `emptyDir` so the data is not removed as long as the Pod is not rescheduled to a different Node. To force download, you need to run:
+
+```bash
+kubectl exec -it $(kubectl get po -l app=botkube -n botkube -oname) -- rm -rf /tmp
+kubectl delete po -l app=botkube -n botkube
+```
+
+Later we can provide a dedicated `@Botkube` command to simplify refreshing downloaded plugins. However, in the happy path scenarios no one should replace already release binaries.
+
+### Supporting multi OS and architectures
+
+We want to run plugins on different Kubernetes distros and also locally for development purposes. As a result, we need to support different operating systems and different architectures.
+
+We enforce that plugin binaries have a suffix with a given pattern: `{os}_{arch}[.exe]`. For example:
+
+- `botkube_executor_helm_linux_amd64`
+- `botkube_source_prometheus_darwin_amd64`
+
+Botkube uses the `runtime.GOOS` and `runtime.GOARCH` variables to determine on which system Botkube is running. It downloads respected binary by searching a proper link in plugin index file:
```

Comment:
> It downloads respected binary by searching a proper link in plugin index file

So it is based on the link itself - looking for specific pattern. Mixed feelings there - we won't be able to support some arbitrary links without such pattern.

WDYT about having it explicitly inside the repository like:

```yaml
entries:
  - name: "kubectl"
    type: "executor"
    description: "Kubectl executor plugin."
    version: "v1.0.0"
    urls: # or `binaries`?
      - url: https://github.com/kubeshop/botkube/releases/download/v0.17.0/executor_kubectl-darwin-amd64
        platform:
          architecture: amd64
          os: linux
     # (...)
```

? This is very similar what's in the OCI image spec: https://github.com/opencontainers/image-spec/blob/main/image-index.md#example-image-index

We can skip `platform` property though and flat it if you like it. But having an object for each `url`/`binary` will be more future-proof if we needed to extend it at some point.

Alternatively, we could have some map, like:

```yaml
entries:
  - name: "kubectl"
    type: "executor"
    description: "Kubectl executor plugin."
    version: "v1.0.0"
    binaries:
      darwin/amd64: 
          url: https://github.com/kubeshop/botkube/releases/download/v0.17.0/executor_kubectl-darwin-amd64
     linux/arm64: { url: "..." }
```

But I like the list more I think 🤔

Reply:
>So it is based on the link itself - looking for specific pattern. Mixed feelings there - we won't be able to support some arbitrary links without such pattern.

I was also thinking about that and I checked how `gh` CLI works and they are doing exactly the same thing. The question is whether there is a real use-case for that. However, I need to agree that the contract is not visible OOTB.

However, I decided to just have it as simple as possible. 

Let's wait also for @huseyinbabal opinion and decide which one to pick 👍

Reply:
what @pkosiec suggested is close to [docker's manifest system](https://docs.docker.com/engine/reference/commandline/manifest/#inspect-an-images-manifest-and-get-the-osarch-info). If we try to use more standardized format, it will help us to detect customer's os/arch with go and map it to exact manifest.

Reply:
FYI: accepted format:
```
  - name: "kubernetes"
    type: "source"
    description: "Kubernetes source plugin."
    version: "v1.0.0"
    urls:
      - url: https://github.com/kubeshop/botkube/releases/download/v0.17.0/source_kubernetes-darwin-amd64
        platform:
          architecture: amd64
          os: darwin
```